message_ix_models/model/water/dsl_engine.py

```diff
@@ -0,0 +1,387 @@
+    df: pd.DataFrame
+        The dataframe to feed the make_df function with.
+    skip_kwargs: list[str]
+        List of keys to skip from the rule.
+    """
+    kwargs = {k: v for k, v in rule.items() if k != "type" and k not in skip_kwargs}
+    dfs_context = {"df": df}  # Wrap the single df in a dict with key 'df'
+    for k, v in kwargs.items():
+        if isinstance(v, str):
+            # Pass the context dictionary to eval_field
+            kwargs[k] = eval_field(v, dfs_context)
+
+    return kwargs
+
+
+def feed_make_df_pl(rule: dict, dfs: dict, skip_kwargs: list[str] = []) -> dict:
+    """Process rule parameters, evaluating DataFrame references and keeping literals.
+
+    Args:
+        rule: Rule parameters dict
+        dfs: DataFrames available for reference (dict mapping name to DataFrame)
+        skip_kwargs: Keys to exclude from processing
+    Returns:
+        Dict of processed parameters
+    """
+    kwargs = {k: v for k, v in rule.items() if k != "type" and k not in skip_kwargs}
+    # df_search_terms = {df_key: f"{df_key}[" for df_key in dfs} # No longer needed
+
+    for k, v in kwargs.items():
+        if isinstance(v, str):
+            # Directly call eval_field, passing the full dfs dictionary.
+            # eval_field will now resolve the correct
+            # dataframes based on prefixes found in v.
+            try:
+                kwargs[k] = eval_field(v, dfs)
+            except KeyError as e:
+                # Re-raise KeyError if eval_field couldn't find a required df prefix
+                raise KeyError(f"Error processing rule key '{k}' with value '{v}': {e}")
+            except Exception as e:
+                # Catch other potential evaluation errors from eval_field
+                raise ValueError(
+                    f"Error evaluating rule key '{k}' with value '{v}': {e}"
+                )
+        # else: # No change needed for non-string values
+        #    kwargs[k] = v
+
+    return kwargs
+
+
+@minimum_version("python 3.10")
+def feed_make_df(
+    rule: dict,
+    rule_dfs: pd.Series | pd.DataFrame | dict,
+    skip_kwargs: list[str] = [],
+) -> dict:
+    """Feed the make_df function with the rule parameters.
+    For each value call the eval_field function to evaluate the value.
+    Parameters
+    ----------
+    rule: dict
+        The rule to feed the make_df function with.
+    rule_dfs: pd.Series | pd.DataFrame | dict
+        The dataframes to feed the make_df function with.
+    skip_kwargs: list[str]
+        List of keys to skip from the rule.
+    """
+    match rule_dfs:
+        case dict():
+            return feed_make_df_pl(rule, rule_dfs, skip_kwargs)
+        case pd.Series() | pd.DataFrame():
+            return feed_make_df_sl(rule, rule_dfs, skip_kwargs)
+        case _:
+            raise ValueError(f"Invalid input type for rule_dfs: {type(rule_dfs)}")
+
+
+@minimum_version("python 3.10")
+def run_standard(
```

Comment:
maybe you could rename this function, as the word "run" has in our context a different meaning. Something like "build" "fill" "parametrize", might be more understandable. not sure